lib/repl.js

```diff
@@ -70,6 +70,88 @@ const BLOCK_SCOPED_ERROR = 'Block-scoped declarations (let, ' +
     'const, function, class) not yet supported outside strict mode';
 
 
+class LineParser {
+
+  constructor() {
+    this.reset();
+  }
+
+  reset() {
+    this._literal = null;
+    this.shouldFail = false;
+    this.blockComment = false;
+  }
+
+  parseLine(line) {
+    var previous = null;
+    this.shouldFail = false;
+    const wasWithinStrLiteral = this._literal !== null;
+
+    for (const current of line) {
+      if (previous === '\\') {
+        // valid escaping, skip processing. previous doesn't matter anymore
+        previous = null;
+        continue;
+      }
+
+      if (!this._literal) {
+        if (previous === '*' && current === '/') {
+          if (this.blockComment) {
+            this.blockComment = false;
+            previous = null;
+            continue;
+          } else {
+            this.shouldFail = true;
+            break;
+          }
+        }
+
+        // ignore rest of the line if `current` and `previous` are `/`s
+        if (previous === current && previous === '/' && !this.blockComment) {
+          break;
+        }
+
+        if (previous === '/' && current === '*') {
+          this.blockComment = true;
+          previous = null;
+        }
+      }
+
+      if (this.blockComment) continue;
+
+      if (current === this._literal) {
+        this._literal = null;
+      } else if (current === '\'' || current === '"') {
+        this._literal = this._literal || current;
+      }
+
+      previous = current;
+    }
+
+    const isWithinStrLiteral = this._literal !== null;
+
+    if (!wasWithinStrLiteral && !isWithinStrLiteral) {
+      // Current line has nothing to do with String literals, trim both ends
+      line = line.trim();
+    } else if (wasWithinStrLiteral && !isWithinStrLiteral) {
+      // was part of a string literal, but it is over now, trim only the end
+      line = line.trimRight();
+    } else if (isWithinStrLiteral && !wasWithinStrLiteral) {
+      // was not part of a string literal, but it is now, trim only the start
+      line = line.trimLeft();
+    }
+
+    const lastChar = line.charAt(line.length - 1);
+
+    this.shouldFail = this.shouldFail ||
+                      ((!this._literal && lastChar === '\\') ||
+                      (this._literal   && lastChar !== '\\'));
+
+    return line;
+  }
```

Comment:
Would this be simpler as a generator ala the readline parser? That way we can pause on comments for more data.